scripts/generators/typescript.js

```diff
@@ -0,0 +1,155 @@
+"use strict";
+
+const t = require("../../packages/babel-types");
+const utils = require("./utils");
+
+const NODE_PREFIX = "";
+
+let code = `// NOTE: This file is autogenerated. Do not modify.
+// See scripts/generators/typescript.js for script used.
+
+export class ${NODE_PREFIX}Comment {
+  value: string;
+  start: number;
+  end: number;
+  loc: ${NODE_PREFIX}SourceLocation;
+}
+
+export class ${NODE_PREFIX}BlockComment extends ${NODE_PREFIX}Comment {
+  type: "BlockComment";
+}
+
+export class ${NODE_PREFIX}LineComment extends ${NODE_PREFIX}Comment {
+  type: "LineComment";
+}
+
+export class ${NODE_PREFIX}SourceLocation {
+  start: {
+    line: number;
+    column: number;
+  };
+
+  end: {
+    line: number;
+    column: number;
+  };
+}
+
+export class Node {
```

Comment:
Same here, but I'd actually like a slightly more complex setup, if possible.

```ts
// these can be accumulated and just written as a declaration at the end of file;
// typescript type definitions have no ordering (think of them as "always hoisted")
type NodeTypes = 'Identifier' | 'BlockStatement' | 'FunctionDeclaration' | ...
type AnyNode = Identifier | BlockStatement | FunctionDeclaration | ...

interface Node<T extends NodeTypes> {
  type: T;
  // NOTE: it is possible to make a node that babel will accept as
  // valid and not have leadingComments on it.
  // Maybe it (and all keys besides `type`) should be marked optional.
    leadingComments: ReadonlyArray<Comment> | null;
  ...
}
```

When a interface for a specific type is generated, say `Identifier`, it should be generated as:

```ts
interface Identifier extends Node<'Identifier'> {
  name: string
}
```

Reply:
That way of writing a declaration is a bit more elaborate; but this means that you are able to use `switch` / `if` or other type narrowing operators to narrow the node type _without_ requiring any type assertions.

```ts
function rename(node: AnyNode, name: string) {
  if (node.type === 'Identifier') {
    // typeof node is Identifier
    return t.clone(node, { name })
  }
  // typeof node is still AnyNode but cannot be Identifier anymore
  if (node.type === 'FunctionDeclaration') {
    return t.cloneDeep(node, { id: t.identifier(name) })
  }
  // typeof node is AnyNode minus Identifier | FunctionDeclaration
  // ...
  return node
} // if t.clone{,Deep} are typed correctly, inferred return type for rename is still AnyNode
```